### ISSUE TITLE
How to perform hessian vector product in a GPyTorch kernel?

### ISSUE THREAD

**Issue by Akella17** [cornellius-gp/gpytorch#831]
I need to design a (Fisher) kernel that is of the format:
k(x<sub>1</sub>, x<sub>2</sub>)  = &nabla;<sub>&theta;</sub>x<sub>1</sub><sup>T</sup> **H** &nabla;<sub>&theta;</sub>x<sub>2</sub>,
Where **H** = &nabla;<sub>&theta;</sub>[&nabla;<sub>&theta;</sub>(KL)], is the hessian of KL divergence.

Since it is computationally intractable to calculate **H** beforehand for the kernel, I need a way to perform hessian vector product on-the-fly during the prediction stage, i.e.,
**H**.v = &nabla;<sub>&theta;</sub>[&nabla;<sub>&theta;</sub>(KL).v]
Similarly, the covariance matrix vector product is as follows:
**K**.v = &nabla;<sub>&theta;</sub>(X<sub>1</sub>.&nabla;<sub>&theta;</sub> (&nabla;<sub>&theta;</sub>(KL).&nabla;<sub>&theta;</sub>(X<sub>2</sub>.v)))

Here, KL is a scalar, v, X1, and X2 are ***t*** dimensional vectors with ***t*** being the number of training examples. All the matrices are **bolded** for easy distinction.

Is there any way I can achieve this with the existing GPyTorch library.

**Comment by jacobrgardner** [comment]
@Akella17 I think this is going to end up being more a PyTorch thing than a GPyTorch thing in general.

PyTorch supports Hessian vector products via two sequential `torch.autograd.grad` calls, and all relevant custom GPyTorch functions are twice differentiable. If you google around, there are a few discussion posts on the pytorch forums about how to do this.

**Comment by Akella17** [comment]
@jacobrgardner I just need a little help locating where exactly the conjugate gradient step over the lazy covariance matrix is happening so that I can create a special case for my custom kernel there.
